### ISSUE TITLE
Feature: Authentication hash

### ISSUE THREAD

**Issue by listerr** [inex/IXP-Manager#269]
I'd like to be able to authenticate logged in IXP Manager users into other web applications so that we can address a few outstanding features we've been asked for which are not present in the IXP Manager portal (yet) 

I'd like to just write all of these things as separate applications to do the $thing, and  integrate it with the same authentication as the portal

My idea is along the lines of a simple authentication cookie/hash that once a user is logged in, can be
set when a user authenticates with the portal so we can seamlessly develop other things around it without having to use the rather cumbersome .htaccess password method.

**Comment by nickhilliard** [comment]
single sign-on is hard.

For non IXP Manager authentication, INEX connects apache up to one of the auth mysql modules, which enables basic http(s) auth.  This is straightforward, works, is language independent, and is easy to set up.  Obviously it has its limitations.

If you're writing php, it's more complicated because you need to drag in a chunk of the framework and handle logins and all that.  @barryo can give a better answer on what exactly you would need to do here.

**Comment by barryo** [comment]
Yeah, this is not particularly easy at all. 

Typical methods would include:

1. Social media type sign on which is effectively OAuth but all apps need to support the OAuth provider. Currently IXP Manager has no distributed OAuth type support and it's not in the plans. However, nor do the other applications you have in mind I guess?
2. Make IXP Manager support login sessions from other applications. Not going to happen to be honest. Once you add in this kind of support (and it's really hacky), you have to maintain and then deal with increasing requests to support more and more third party applications.
3. Make your other applications support IXP Manager's sessions (the converse of (2) above). This really boils down to:
   * cookie support for portal.lonap.net (or *.lonap.net) in your other apps. I.e. the scope of the IXP Manager cookie / URL of your other apps must be in scope
   * code your other application to read the IXP Manager cookie and fetch the login session details from the session backend
   * there is a requirement here to have shared session storage / accessible session storage (e.g. MySQL session support in PHP)
   * if your other application is not PHP based, it would need to be able to unserialize PHP objects 
   * there's a few other requirements / gotchas I can go into if this is really a runner but... yuck

Note that (3) is what would be required in IXP Manager in (2) if we were going to go down that road. You can see why we wouldn't...

Note that all of the above is about single sign on rather than shared credentials. You could still use the IXP Manager database for shared credentials as well as the above but these are two separate things. If you're not using shared credentials, then your code in the other apps would need to be able to auto-create users if appropriate or else some other kind of scripted user synchronisation. 

The bottom line is that SSO requires application (programatic) support and is not something that's on our published development plan. There may also be future potential issues such as Laravel potentially encrypting cookie data. It'd be symmetric key encryption but it's another potential future layer of complication.

Nick has explained how we do it at INEX but, as you know, this requires a second htaccess-type step.

Sorry @listerr - not the easy answer you were hoping for 😞

**Comment by listerr** [comment]
Thanks for the explanation, Barry.

The reason behind asking is that there are many things that I could
probably implement myself that are not in IXP Manager's interface, and,
well, we just need to get on with a few things. :)

So it may well be worth implementing a simple authentication cookie /
hash system that other CGIs can read to confirm the current user and
login state.

This would allow me to implement our own things that we've been wanting
for some time, but are unable to implement in IXP manager directly, for
example, we would like to implement simple web forms in the right places
for example for registering to attend the meeting without either having
to re-key all the information again that's in the portal, or coming
unstuck when a .htaccess password box pops up.

Ordering / modifying services e.g, upgrade / cancel / change this port.

We need to improve our process for upgrading port and the ordering
process etc, and we need some simple forms to make this happen.

For years we've been having conversations of:

"Hmm. This process went wrong for that customer. We should have a proper
form for this. Let's make a form so we can check the details."

"Oh, we can't because reasons in IXP Manager."

"We can make a form anyway, but then they have to type it all in again
anyway."

"Yeah, that's no so good."

It's the classic interface case of: "You already have all of this
information. Why are you asking for it all over again?" scenario.

So it may well be worth it, so that you can say "This is not currently
supported in IXP Manager, however, we do support an authentication
system that allows you to easily integrate your own thing with it."

Or, we will just keep asking again and again for things to be
implemented, or fixed in a way that works for us.

- Members being able to change things like rDNS (Members are surprised
to lean that they can't do this already. We a lot of requests for
this. Would like to write something to make this go away.)

- Custom implementation for our mailing lists, because the current IXP
manager thing doesn't work for us (has been broken for some time,
and other issues there with it when it did work.)

- Surveys, other forms requiring feedback or data to be filled in etc.

- Links to, for example documents or monitoring systems.

So, this is about writing our own stuff and hooking it in with IXP
Manager in a seamless way. (or as seamless as we can make it.)

Seems many password managers do not support http basic auth, so this is
my least favourite option, asking for another password/login, even if it
is the same.

So I'd like a way to make these long standing issues go away for our
users, so we can become more self-sufficient, rather than the current
posision of "It's not in IXP Manager. Not going to happen any time soon,
if ever. We can never implement this thing for our users."

So it could be worth your effort, then we can make some progress and get
out of your hair with all these silly feature requests that we want.



Regards,


Rob


-- 
Rob Lister
rob@lonap.net
+44 20 3137 8330
